docker-images/kafka/scripts/kafka_liveness.sh

```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+if [ -f /var/opt/kafka/kafka-ready ] ; then
+  rm /var/opt/kafka/zk-connected 2&> /dev/null
+  # Test listening on replication port 9091
+  netstat -lnt | grep -Eq 'tcp6?[[:space:]]+[0-9]+[[:space:]]+[0-9]+[[:space:]]+[^ ]+:9091.*LISTEN[[:space:]]*'
+else
+  # Not yet ready, so test ZK connected state
+  test -f /var/opt/kafka/zk-connected
```

Comment:
@tombentley maybe a dump question ... let's imagine that a broker can connect to ZK so this file is created but it never goes in the running state so the kafka-ready is never created. Does it mean that this liveness will always return true (because the ZK connected file exists) but actually the broker "is not live" and it will be never restarted by Kube (which could be a fix for having the broker starting fine)?

Reply:
That's right, but how would you propose to distinguish that case from the case where the broker is just taking a long time to become ready? If there's large log to recover it could take a while, and having kube kill those pods because we decided they weren't live would be the wrong thing to do (because the restarted pod would also have to do log recovery).

We could spend a lot of time coming up with increasingly complicated heuristics for determining whether a broker is really alive (e.g. if before ready then is the log recovery actually progressing? If after ready then can we produce and/or consume). There would always be edge cases where the probe was giving the wrong answer though. So we need a heuristic which is good enough.

Reply:
Good point. So it means that in case of real problems (so not a broker which is ok but it just takes long time to start), the user has to notice that a kill the pod manually for restarting the broker?

Reply:
Yes, more or less. The pod would be in the Unready state though, so I think it would be relatively easy to identify. And the onus is then on the user to look at the logs and decide whether this requires patience or a broker restart.